text/3778-inline-intents.md

```diff
@@ -0,0 +1,189 @@
+
+In most cases, plain `#[inline]` is fine, especially with PGO and LTO.
+However, if you've measured and things are making a poor choice, there are some
+options you can use to hint the compiler towards what you want.
+
+## `inline(trampoline)`
+
+This is intended for functions which quickly "bounce" the caller off to some
+other implementation, after doing some initial checks or transformations.
+
+For example, this is useful in a safe function which does some safety checks,
+then calls an `unsafe` version of the function to do the actual work.  Or maybe
+it's a function with a common trivial path, but which sometimes needs to call
+out to a more complicated version, like how `Vec::push` is usually trivial but
+occasionally needs to reallocate.
+
+## `inline(rarely)`
+
+This is intended for functions which normally shouldn't be inlined, but where
+exceptions exist so you don't want full `inline(never)`.
+
+For example, maybe this is a vectorized loop that you wouldn't want copied into
+every caller, but you know that using it for short arrays is common, and thus
+want the back-end to be able to inline and fully unroll it in those cases.
+
+## In combination
+
+These can work particularly well together.
+
+For example, perhaps the public function is `inline(trampoline)` so that the
+initial `NonZero::new` check can be inlined into the caller (where it's more
+easily optimized out) but that calls a private `inline(rarely)` function which
+takes `NonZero` to avoid needing extra checks internally.
+
+Or perhaps you write an `inline(trampoline)` function that picks a strategy
+based on the types and arguments, then dispatches to one of many possible
+`inline(rarely)` implementations.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Like `inline`, these "do nothing" in a spec sense.
+
+The only language change is thus in allowing the two new tokens in the attribute
+(in addition to `never` and `always`).
+
+## Implementation options
+
+⚠ These are not normative, just to illustrate possibilities. ⚠
+
+In LLVM, `#[inline]` sets the [`inlinehint` function attribute](https://llvm.org/docs/LangRef.html#function-attributes),
+so `inline(rarely)` could skip doing that, and thus comparatively slightly
+discourage inlining it.
+
+In MIR inlining, we already attempt to deduce trampolines as of [#127113].
+This would let people opt-in to that behaviour, even in places it's not obvious.
+Then we could allow inlining trampolines into trampolines, but avoid inlining
+non-trampolines into a trampoline.  And we have an internal attribute
+`rustc_no_mir_inline` which blocks MIR inlining, so `inline(rarely)` could also
+do that (or maybe just make the threshold very restrictive).
+
+[#127113]: https://github.com/rust-lang/rust/pull/127113
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+These are still up to the programmer to get right, so
+- they might just make analysis paralysis worse
```

Comment:
There's also `inline(usually)` (https://github.com/rust-lang/rust/pull/130679) which I think is very well-motivated but adds to this problem. It would be great if we could just make the user-facing `inline(always)` work that way and keep an internal attribute like `#[rustc_force_inline]` for the cases (intrinsics) that need inlining even in opt-level=0 builds, but it's not clear if that will work out in the end.

Reply:
Just to keep conversation up to date, `#[rustc_force_inline]` has landed and the intent is to use it to implement intrinsics: https://github.com/rust-lang/rust/pull/134082.